src/coreclr/jit/lsrabuild.cpp

```diff
@@ -3968,6 +3998,12 @@ int LinearScan::BuildReturn(GenTree* tree)
                             {
                                 buildInternalIntRegisterDefForNode(tree, dstRegMask);
                             }
+#if defined(TARGET_XARCH) && defined(FEATURE_SIMD)
+                            else if (varTypeUsesMaskReg(dstType))
```

Comment:
Should this check be guarded for AVX512 platforms only to reduce the impact of TP on non-AVX512 xarch platforms?

Reply:
`varTypeUsesMaskReg` is basically doing a  `dstType == TYP_MASK` check (and we could actually optimize it to that given we only have the 1 mask type today).

Checking `comp->canUseEvexEncoding` would just add an additional branch and not really save us much (we'd be trading one comparison/branch for another). It may even and potentially slow down the actual execution due to more branches being in a small code window which can pessimize the branch predictor.

Reply:
sure, that's what I thought, because it is still a runtime check and will cost us. In LSRA, I think this check would have increased the TP regression.

Reply:
Right. It will increase the cost to get a floating-point internal register. But, that cost basically "has" to exist; we're either going to have `else if (comp->canUseEvexEncoding && varTypeUsesMaskReg(dstType))` -or- `else if (varTypeUsesMaskReg(dstType))` -or- `else if (varTypeUsesFloatReg(dstType))`

The first one is a `memory access`, `branch`, `comparison`, and `branch`

The second one is currently just a `comparison` and `branch`.

The third is currently a `memory access`, `comparison`, and `branch`

So what we have currently is the cheapest we can make it.

Reply:
We could potentially have a small lookup table that maps `type` to a `fnptr` representing the `buildInternalReg` call. However, that's potentially more expensive in terms of actual execution cost.

I'd think there are better ways to win the TP back instead.